clang/test/Preprocessor/sycl-macro-target-specific.cpp

```diff
@@ -1,3 +1,16 @@
+// RUN: %clang_cc1 %s -fsycl-is-device -triple nvptx64-nvidia-nvcl-sycldevice -E -dM \
+// RUN: | FileCheck --check-prefix=CHECK-NVPTX %s
+// RUN: %clang_cc1 %s -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -E -dM \
+// RUN: | FileCheck --check-prefix=CHECK-NVPTX-NEG %s
+// RUN: %clang_cc1 %s -fsycl-is-device -triple spir64_gen-unknown-unknown-sycldevice -E -dM \
+// RUN: | FileCheck --check-prefix=CHECK-NVPTX-NEG %s
+// RUN: %clang_cc1 %s -fsycl-is-device -triple spir64_x86_64-unknown-unknown-sycldevice -E -dM \
+// RUN: | FileCheck --check-prefix=CHECK-NVPTX-NEG %s
+// RUN: %clang_cc1 %s -fsycl-is-device -triple spir64_fpga-unknown-unknown-sycldevice -E -dM \
+// RUN: | FileCheck --check-prefix=CHECK-NVPTX-NEG %s
+// CHECK-NVPTX: #define __SYCL_NVPTX__
+// CHECK-NVPTX-NEG-NOT: #define __SYCL_NVPTX__
```

Comment:
> Can't use we `__SYCL_DEVICE_ONLY__` and `__NVPTX__` instead of `__SYCL_NVPTX__`?

You could. They are used in conjunction in both `sycl/include/CL/sycl/ONEAPI/sub_group.hpp` and `sycl/include/CL/sycl/detail/spirv.hpp`.

Reply:
Can we drop setting `__SYCL_NVPTX__` and replace all uses with combination of `__SYCL_DEVICE_ONLY__` and `__NVPTX__`?

Reply:
Only problem I could see with doing this is that sycl/doc/CompilerAndRuntimeDesign.md suggests the use of `__SYCL_NVPTX__` when using NVPTX builtins, so it may be in use elsewhere.

Reply:
The only uses I'm able to find are in `sycl/include/CL/__spirv/spirv_vars.hpp`, but they seems to be safe to replace with `__NVPTX__`.

Reply:
What I mean is there may be user code out there using it to access NVPTX functionality, though it is purely speculation.

Reply:
Do I get it correctly that `__NVPTX__` gets defined automatically when the `nvptx64` triple is used? (Looking at `Builder.defineMacro("__NVPTX__")` in `calng/lib/Basic/Targets/NVPTX.cpp`)

If that is the case, we could maybe drop `__SYCL_NVPTX__` from FE, but define it under `#if defined(__NVPTX__) && defined(__SYCL_DEVICE_ONLY__)` to provide backwards compatibility for the (hypothetical) custom NVPTX builtin uses.

Reply:
> 
> 
> Do I get it correctly that `__NVPTX__` gets defined automatically when the `nvptx64` triple is used? (Looking at `Builder.defineMacro("__NVPTX__")` in `clang/lib/Basic/Targets/NVPTX.cpp`)]

To my knowledge, yes.

> If that is the case, we could maybe drop `__SYCL_NVPTX__` from FE, but define it in the headers under `#if defined(__NVPTX__) && defined(__SYCL_DEVICE_ONLY__)` to provide backwards compatibility for the (hypothetical) custom NVPTX builtin uses.

I am okay with this solution. 👍

Reply:
> What I mean is there may be user code out there using it to access NVPTX functionality, though it is purely speculation.

IMO, CompilerAndRuntimeDesign.md is not indented to be a user guide. It describes internal design for project contributors and we should be free to change it as long as user defined behavior is not impacted.

> Do I get it correctly that `__NVPTX__` gets defined automatically when the `nvptx64` triple is used? (Looking at `Builder.defineMacro("__NVPTX__")` in `clang/lib/Basic/Targets/NVPTX.cpp`)

I think so.

> If that is the case, we could maybe drop `__SYCL_NVPTX__` from FE, but define it in the headers under `#if defined(__NVPTX__) && defined(__SYCL_DEVICE_ONLY__)` to provide backwards compatibility for the (hypothetical) custom NVPTX builtin uses.

Defining it in the headers won't work unless headers are included, so I don't think it's necessary. Users can easily create a replacement for `__SYCL_NVPTX__` if they are using it for some reason and not willing to replace with `#if defined(__NVPTX__) && defined(__SYCL_DEVICE_ONLY__)`.

I suggest remove setting `__SYCL_NVPTX__` to avoid maintenance burden and use already supported alternatives. We won't be able to upstream this code without proper justification.

Reply:
> IMO, CompilerAndRuntimeDesign.md is not indented to be a user guide. It describes internal design for project contributors and we should be free to change it as long as user defined behavior is not impacted.

Though I agree, it has previously been the only way to access select NVPTX features, so it is not impossible that someone out there could be using it. That said, if removing it is preferred I rather it be done sooner rather than later.

Reply:
Sounds good to me, I'll try implementing this within the current PR - the work doesn't seem extensive and is going to involve FE changes anyway.

Reply:
> Sounds good to me, I'll try implementing this within the current PR - the work doesn't seem extensive and is going to involve FE changes anyway.

Right. I see only 3 files mentioning this macro, so it shouldn't be hard to do.
Thanks!